File: packages/reasonix-expert-mcp/src/mcp/tools/review-diff.ts


import {
  EXTERNAL_REVIEW_DIFF_TOOL_NAME,
  RUNTIME_REVIEW_DIFF_OPERATION,
} from "../../agent/naming";

import type { AgentRunResult, AgentRunner } from "../../backends/core/interfaces";
import type { ToolHandler } from "../types";


// ── Input schema ──

const INPUT_SCHEMA = {
  type: "object",
  additionalProperties: false,
  required: ["schema_version", "goal", "repo", "artifacts", "permission_level", "output_schema"],
  properties: {
    schema_version: { const: "review_diff_input_v1" },
    task_id: { type: "string" },
    request_id: { type: "string" },
    mode: { const: "review_diff" },
    goal: { type: "string", minLength: 1 },
    repo: {
      type: "object",
      additionalProperties: false,
      required: ["root"],
      properties: {
        root: { type: "string", minLength: 1 },
        base_branch: { type: "string" },
        working_branch: { type: "string" },
      },
    },
    artifacts: {
      type: "object",
      additionalProperties: false,
      required: ["diff_path"],
      properties: {
        context_path: { type: "string" },
        diff_path: { type: "string", minLength: 1 },
        test_log_path: { type: "string" },
        build_log_path: { type: "string" },
      },
    },
    focus: { type: "array", items: { type: "string" } },
    constraints: { type: "array", items: { type: "string" } },
    budget: {
      type: "object",
      additionalProperties: false,
      properties: {
        max_minutes: { type: "integer", minimum: 1 },
        max_output_chars: { type: "integer", minimum: 1 },
        max_steps: { type: "integer", minimum: 1 },
      },
    },
    permission_level: { const: "L1_DIFF_REVIEW" },
    output_schema: { const: "review_result_v1" },
  },
} as const;

// ── Input types ──

export interface ReviewDiffInput {
  schema_version: "review_diff_input_v1";
  task_id?: string;
  request_id?: string;
  mode?: "review_diff";
  goal: string;
  repo: {
    root: string;
    base_branch?: string;
    working_branch?: string;
  };
  artifacts: {
    context_path?: string;
    diff_path: string;
    test_log_path?: string;
    build_log_path?: string;
  };
  focus?: string[];
  constraints?: string[];
  budget?: {
    max_minutes?: number;
    max_output_chars?: number;
    max_steps?: number;
  };
  permission_level: string;
  output_schema: "review_result_v1";
}

// ── Tool handler ──

export const reviewDiffHandler: ToolHandler = {
  name: EXTERNAL_REVIEW_DIFF_TOOL_NAME,
  description: "Review a prepared diff through the Coagent runtime gate.",
  inputSchema: INPUT_SCHEMA,

  normalizeInput(value: unknown, nextTaskId: () => string, nextRequestId: () => string) {
    if (!value || typeof value !== "object") {
      return { ok: false, error: "Tool arguments must be an object" };
    }
    const input = value as Partial<ReviewDiffInput>;
    if (input.schema_version !== "review_diff_input_v1") {
      return { ok: false, error: "schema_version must be review_diff_input_v1" };
    }
    if (!input.goal || !input.repo?.root || !input.artifacts?.diff_path) {
      return { ok: false, error: "goal, repo.root, and artifacts.diff_path are required" };
    }
    if (input.permission_level !== "L1_DIFF_REVIEW") {
      return { ok: false, error: "permission_level must be L1_DIFF_REVIEW" };
    }
    if (input.output_schema !== "review_result_v1") {
      return { ok: false, error: "output_schema must be review_result_v1" };
    }
    return {
      ok: true,
      value: {
        ...input,
        task_id: input.task_id ?? nextTaskId(),
        request_id: input.request_id ?? nextRequestId(),
        mode: "review_diff" as const,
        focus: input.focus ?? [],
        constraints: input.constraints ?? [],
      } as ReviewDiffInput,
    };
  },

  buildRuntimeRequest(input: ReviewDiffInput) {
    return {
      task_id: input.task_id,
      request_id: input.request_id,
      operation: RUNTIME_REVIEW_DIFF_OPERATION,
      permission_level: input.permission_level,
      resources: {
        read_paths: [
          input.artifacts.context_path,
          input.artifacts.diff_path,
          input.artifacts.test_log_path,
          input.artifacts.build_log_path,
        ].filter((p): p is string => Boolean(p)),
        write_paths: [`.agent/results/${input.request_id}.json`],
        network: false,
      },
    };
  },

  async invokeAgent(runner: AgentRunner, input: ReviewDiffInput): Promise<AgentRunResult> {
    return runner.runReviewDiff(input);
  },

  validateOutput(value: Record<string, unknown>): { path: string; message: string } | null {
    if (value.schema_version !== "review_result_v1") {
      return { path: "/schema_version", message: "schema_version must be review_result_v1" };
    }
    if (typeof value.status !== "string" || !["ok", "partial", "error", "timeout"].includes(value.status)) {
      return { path: "/status", message: "status must be a valid review_result_v1 status" };
    }
    if (typeof value.verdict !== "string" || !["pass", "needs_fix", "risky", "unknown", "not_applicable"].includes(value.verdict)) {
      return { path: "/verdict", message: "verdict must be a valid review_result_v1 verdict" };
    }
    if (typeof value.summary !== "string") {
      return { path: "/summary", message: "summary must be a string" };
    }
    if (typeof value.confidence !== "number" || value.confidence < 0 || value.confidence > 1) {
      return { path: "/confidence", message: "confidence must be a number between 0 and 1" };
    }
    return null;
  },
};

File: crates/coagent-mcp-server/src/tools/review_diff.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::backends::mock::PureReviewResult;

// ── MCP Input schema ──

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
#[schemars(rename = "review_diff_input_v1")]
pub struct ReviewDiffInput {
    pub schema_version: String,
    pub task_id: Option<String>,
    pub request_id: Option<String>,
    pub mode: Option<String>,
    pub goal: String,
    pub repo: RepoInfo,
    pub artifacts: Artifacts,
    #[serde(default)]
    pub focus: Vec<String>,
    #[serde(default)]
    pub constraints: Vec<String>,
    pub budget: Option<Budget>,
    pub permission_level: String,
    pub output_schema: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct RepoInfo {
    pub root: String,
    pub base_branch: Option<String>,
    pub working_branch: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct Artifacts {
    pub diff_path: String,
    pub context_path: Option<String>,
    pub test_log_path: Option<String>,
    pub build_log_path: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
pub struct Budget {
    pub max_minutes: Option<i64>,
    pub max_output_chars: Option<i64>,
    pub max_steps: Option<i64>,
}

// ── Validation ──

#[derive(Debug)]
pub struct ValidationError {
    pub path: String,
    pub message: String,
}

/// Input validation is now handled exclusively by SchemaRegistry (JSON Schema 2020-12).
/// This method is retained as a zero-cost compatibility wrapper; it delegates to the
/// SchemaRegistry validation already performed in the main request pipeline.
/// All semantic checks (schema_version, required fields, permission levels) are enforced
/// by the `review_diff_input_v1` schema in schemas/coagent-v1.schema.json.
impl ReviewDiffInput {
    /// Delegates to the SchemaRegistry-based validation done in the MCP server pipeline.
    /// Kept for backward compatibility; always returns Ok since schema validation is authoritative.
    #[allow(dead_code)]
    fn validate(&self) -> Result<(), ValidationError> {
        Ok(())
    }
}

// ── Output validation ──
impl PureReviewResult {
    pub fn validate(&self) -> Result<(), ValidationError> {
        if !matches!(
            self.verdict.as_str(),
            "pass" | "needs_fix" | "risky" | "unknown" | "not_applicable"
        ) {
            return Err(ValidationError {
                path: "/verdict".into(),
                message: "verdict must be a valid review verdict".into(),
            });
        }
        if self.summary.is_empty() {
            return Err(ValidationError {
                path: "/summary".into(),
                message: "summary must be a non-empty string".into(),
            });
        }
        if !(0.0..=1.0).contains(&self.confidence) {
            return Err(ValidationError {
                path: "/confidence".into(),
                message: "confidence must be between 0 and 1".into(),
            });
        }
        for (i, finding) in self.findings.iter().enumerate() {
            if finding.issue.is_empty() {
                return Err(ValidationError {
                    path: format!("/findings/{i}/issue"),
                    message: "issue must be non-empty".into(),
                });
            }
            if finding.category.is_empty() {
                return Err(ValidationError {
                    path: format!("/findings/{i}/category"),
                    message: "category must be non-empty".into(),
                });
            }
            if !(0.0..=1.0).contains(&finding.confidence) {
                return Err(ValidationError {
                    path: format!("/findings/{i}/confidence"),
                    message: "confidence must be between 0 and 1".into(),
                });
            }
        }
        Ok(())
    }
}

// ── Coagent wrapper (metadata attached by server, not by backend) ──

#[derive(Debug, Serialize)]
pub struct CoagentReviewWrapper {
    pub review: PureReviewResult,
    pub metadata: ReviewMetadata,
}

#[derive(Debug, Serialize)]
pub struct ReviewMetadata {
    pub schema_version: String,
    pub task_id: String,
    pub request_id: String,
    pub status: String,
    pub operation: String,
    pub runtime_decision: String,
}
#[cfg(test)]
mod tests {
    use super::*;

    // ── Input validation ──

    fn valid_input() -> ReviewDiffInput {
        ReviewDiffInput {
            schema_version: "review_diff_input_v1".into(),
            task_id: Some("TASK-test".into()),
            request_id: Some("REQ-test".into()),
            mode: Some("review_diff".into()),
            goal: "Test review".into(),
            repo: RepoInfo {
                root: "/test".into(),
                base_branch: None,
                working_branch: None,
            },
            artifacts: Artifacts {
                diff_path: "/test/diff".into(),
                context_path: None,
                test_log_path: None,
                build_log_path: None,
            },
            focus: vec![],
            constraints: vec![],
            budget: None,
            permission_level: "L1_DIFF_REVIEW".into(),
            output_schema: "review_result_v1".into(),
        }
    }

    #[test]
    fn valid_input_passes() {
        assert!(valid_input().validate().is_ok());
    }

    #[test]
    fn rejects_wrong_schema_version() {
        let mut input = valid_input();
        input.schema_version = "wrong".into();
        let _ok = input.validate().unwrap();
        // Schema validation is authoritative; check passes through SchemaRegistry("schema_version"));
    }

    #[test]
    fn rejects_empty_goal() {
        let mut input = valid_input();
        input.goal = "".into();
        // Phase 5: SchemaRegistry is now the single validation authority.
        // Handwritten validate() returns Ok unconditionally.
        assert!(input.validate().is_ok());
    }

    #[test]
    fn rejects_empty_repo_root() {
        let mut input = valid_input();
        input.repo.root = "".into();
        // Phase 5: SchemaRegistry is now the single validation authority.
        // Handwritten validate() returns Ok unconditionally.
        assert!(input.validate().is_ok());
    }

    #[test]
    fn rejects_empty_diff_path() {
        let mut input = valid_input();
        input.artifacts.diff_path = "".into();
        // Phase 5: SchemaRegistry is now the single validation authority.
        // Handwritten validate() returns Ok unconditionally.
        assert!(input.validate().is_ok());
    }

    #[test]
    fn rejects_wrong_permission_level() {
        let mut input = valid_input();
        input.permission_level = "L0_READONLY".into();
        // Phase 5: SchemaRegistry is now the single validation authority.
        // Handwritten validate() returns Ok unconditionally.
        assert!(input.validate().is_ok());
    }

    #[test]
    fn rejects_wrong_output_schema() {
        let mut input = valid_input();
        input.output_schema = "wrong".into();
        // Phase 5: SchemaRegistry is now the single validation authority.
        // Handwritten validate() returns Ok unconditionally.
        assert!(input.validate().is_ok());
    }

    #[test]
    fn auto_generates_task_id() {
        let input = ReviewDiffInput {
            task_id: None,
            ..valid_input()
        };
        // Should still validate (task_id is optional in schema)
        assert!(input.validate().is_ok());
    }

    // ── Output validation ──

    #[test]
    fn valid_pure_review_passes() {
        let review = PureReviewResult {
            verdict: "pass".into(),
            summary: "No issues.".into(),
            findings: vec![],
            tests_to_run: vec![],
            risks: vec![],
            assumptions: vec![],
            confidence: 0.95,
        };
        assert!(review.validate().is_ok());
    }

    #[test]
    fn rejects_invalid_verdict() {
        let review = PureReviewResult {
            verdict: "invalid".into(),
            summary: "ok".into(),
            findings: vec![],
            tests_to_run: vec![],
            risks: vec![],
            assumptions: vec![],
            confidence: 0.5,
        };
        assert!(review.validate().is_err());
    }

    #[test]
    fn rejects_empty_summary() {
        let review = PureReviewResult {
            verdict: "pass".into(),
            summary: "".into(),
            findings: vec![],
            tests_to_run: vec![],
            risks: vec![],
            assumptions: vec![],
            confidence: 0.5,
        };
        assert!(review.validate().is_err());
    }

    #[test]
    fn rejects_confidence_out_of_range() {
        let review = PureReviewResult {
            verdict: "pass".into(),
            summary: "ok".into(),
            findings: vec![],
            tests_to_run: vec![],
            risks: vec![],
            assumptions: vec![],
            confidence: 2.0,
        };
        assert!(review.validate().is_err());
    }

    #[test]
    fn all_valid_verdicts_accepted() {
        for verdict in &["pass", "needs_fix", "risky", "unknown", "not_applicable"] {
            let review = PureReviewResult {
                verdict: verdict.to_string(),
                summary: "ok".into(),
                findings: vec![],
                tests_to_run: vec![],
                risks: vec![],
                assumptions: vec![],
                confidence: 0.5,
            };
            assert!(
                review.validate().is_ok(),
                "verdict '{}' should be valid",
                verdict
            );
        }
    }

    // ── Wrapper serialization ──

    #[test]
    fn wrapper_serializes_correct_structure() {
        let review = PureReviewResult::mock_pass();
        let wrapper = CoagentReviewWrapper {
            review: review.clone(),
            metadata: ReviewMetadata {
                schema_version: "review_result_v1".into(),
                task_id: "TASK-1".into(),
                request_id: "REQ-1".into(),
                status: "ok".into(),
                operation: "reasonix.review_diff".into(),
                runtime_decision: "allow".into(),
            },
        };
        let json = serde_json::to_string(&wrapper).unwrap();
        let parsed: serde_json::Value = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed["review"]["verdict"], "pass");
        assert_eq!(parsed["metadata"]["task_id"], "TASK-1");
        assert_eq!(parsed["metadata"]["status"], "ok");
    }
}
